.orchestra/support/distributable-test.py

```diff
@@ -0,0 +1,188 @@
+            env={k: v for k, v in os.environ.items() if k not in FILTER_ENV},
+            stdout=out,
+            stderr=out,
+            text=True,
+        )
+        port = get_listen_port(process.pid)
+        address = f"127.0.0.1:{port}"
+    else:
+        raise ValueError(f"Unknown mode: {sys.argv[2]}")
+
+    try:
+        check_server_up(address)
+    except ValueError as e:
+        print_fd(out_fd)
+        raise e
+
+    transport = RequestsHTTPTransport(f"http://{address}/graphql/")
+    client = Client(transport=transport, fetch_schema_from_transport=True)
+
+    try:
+        print("### RUNNING DISTRIBUTABLE E2E TEST ###")
+        test_revng_distributable(client)
+    except Exception as e:
+        traceback.print_exception(e.__class__, e, e.__traceback__)
+        process.communicate()
+        print_fd(out_fd)
+        sys.exit(1)
+
+    process.terminate()
+    return_code = process.wait()
+    if return_code not in (0, -15):
+        print(f"Daemon exited with code {return_code}")
+        print_fd(out_fd)
+        sys.exit(1)
+
+
+def test_revng_distributable(client):
+    upload_q = gql(
+        """
+        mutation upload($file: Upload!) {
+            uploadFile(file: $file, container: "input")
+        }
+    """
+    )
+    with open(sys.argv[1], "rb") as binary_file:
+        client.execute(upload_q, variable_values={"file": binary_file}, upload_files=True)
+    print("Upload complete")
+
+    client.execute(gql("mutation { runAllAnalyses }"))
+    print("Autoanalysis complete")
+
+    q = gql(
+        """
+    {
+      info {
+        steps {
+          name
+          parent
+          artifacts {
+            kind {
+              name
+            }
+            container {
+              name
+            }
+          }
+        }
+      }
+    }
+    """
+    )
+
+    result = client.execute(q)
+
+    steps = {}
+    topo_sorter = TopologicalSorter()
+    for step in result["info"]["steps"]:
+        steps[step["name"]] = step
+        if step["parent"] is not None:
+            topo_sorter.add(step["name"], step["parent"])
+
+    for step_name in topo_sorter.static_order():
+        if steps[step_name]["artifacts"] is None:
+            continue
+
+        artifacts_container = steps[step_name]["artifacts"]["container"]["name"]
+        artifacts_kind = steps[step_name]["artifacts"]["kind"]["name"]
+
+        q = gql(
+            """
+        query cq($step: String!, $container: String!) {
+          container(name: $container, step: $step) {
+            targets {
+              serialized
+            }
+          }
+        }
+        """
+        )
+        res = client.execute(q, {"step": step_name, "container": artifacts_container})
+
+        for target in res["container"]["targets"]:
+            if not target["serialized"].endswith(f":{artifacts_kind}"):
+                continue
+
+            print(f"Producing {step_name}/{artifacts_container}/{target['serialized']}")
```

Comment:
If it's debug information, make `log` function and write to `stderr`.

Reply:
Done